backend/src/v2/compiler/argocompiler/argo.go

```diff
@@ -180,69 +184,96 @@ func (c *workflowCompiler) templateName(componentName string) string {
 	return componentName
 }
 
-// WIP: store component spec, task spec and executor spec in annotations
-
 const (
-	annotationComponents     = "pipelines.kubeflow.org/components-"
-	annotationContainers     = "pipelines.kubeflow.org/implementations-"
-	annotationKubernetesSpec = "pipelines.kubeflow.org/kubernetes-"
+	argumentsComponents     = "components-"
+	argumentsContainers     = "implementations-"
+	argumentsKubernetesSpec = "kubernetes-"
 )
 
 func (c *workflowCompiler) saveComponentSpec(name string, spec *pipelinespec.ComponentSpec) error {
-	return c.saveProtoToAnnotation(annotationComponents+name, spec)
+	baseComponentName := ExtractBaseComponentName(argumentsComponents + name)
+	return c.saveProtoToArguments(baseComponentName, spec)
 }
 
 // useComponentSpec returns a placeholder we can refer to the component spec
 // in argo workflow fields.
 func (c *workflowCompiler) useComponentSpec(name string) (string, error) {
-	return c.annotationPlaceholder(annotationComponents + name)
+	baseComponentName := ExtractBaseComponentName(argumentsComponents + name)
+	return c.argumentsPlaceholder(baseComponentName)
 }
 
 func (c *workflowCompiler) saveComponentImpl(name string, msg proto.Message) error {
-	return c.saveProtoToAnnotation(annotationContainers+name, msg)
+	baseComponentName := ExtractBaseComponentName(argumentsContainers + name)
+	return c.saveProtoToArguments(baseComponentName, msg)
 }
 
 func (c *workflowCompiler) useComponentImpl(name string) (string, error) {
-	return c.annotationPlaceholder(annotationContainers + name)
+	baseComponentName := ExtractBaseComponentName(argumentsContainers + name)
+	return c.argumentsPlaceholder(baseComponentName)
 }
 
 func (c *workflowCompiler) saveKubernetesSpec(name string, spec *structpb.Struct) error {
-	return c.saveProtoToAnnotation(annotationKubernetesSpec+name, spec)
+	return c.saveProtoToArguments(argumentsKubernetesSpec+name, spec)
 }
 
 func (c *workflowCompiler) useKubernetesImpl(name string) (string, error) {
-	return c.annotationPlaceholder(annotationKubernetesSpec + name)
+	return c.argumentsPlaceholder(argumentsKubernetesSpec + name)
 }
```

Comment:
was there a reason not to deduplicate here? 

I would think kubernetesSpec benefit from this in the same way:

e.g: 

```python
    for i in range(2):
        task = bar(name="bar")
        kubernetes.use_secret_as_env(
            task,
            secret_name='aws-s3-creds',
            secret_key_to_env={'AWS_SECRET_ACCESS_KEY': 'AWS_SECRET_ACCESS_KEY'})
```

Reply:
`kubernetesSpec` can vary across multiple instances of a given component, e.g.:

```python
from kfp import dsl, kubernetes

@dsl.component
def foo():
    print("foo")

@dsl.pipeline
def pipeline():
    task1 = foo()
    task2 = foo()
    kubernetes.add_pod_annotation(
        task1,
        annotation_key='run_id',
        annotation_value='123456',
    )
```

In the above example, `task1` will have a distinct `kubernetesSpec` vis a vis `task2`. Deduplicating the `kubernetesSpec` will fail to account for those distinctions.

Caveat about the above: **...I think** 🙃 .

Really appreciate the thorough review, @HumairAK!

Reply:
very good point! makes sense to me